### PR TITLE
 Issue 1316: A bookie with non-writable dirs should be able to start in readonly mode

### PR DESCRIPTION
**Pull request by sijie**
Descriptions of the changes in this PR:

*Problem*

Bookie failed to start when it doesn't have non-writable dirs. Issue is reported at #1316 

*Solution*

- Introduce a setting `minUsageSizeForEntryLogCreation` to allow creating entry logs even when there is no writable dirs. This setting is replacing `getIsForceGCAllowWhenNoSpace` for creating new log, since entry log files are created not only for gc/compaction. It can also happen on startup and also journal replays.

- Defer creating entry log files until first write happens, when there is no writable ledger dirs.

- add bunch of tests for EntryLog and Bookie initialization.

Master Issue: #1316 
Related Issue: #190

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue #>: Description of pull request`
>     `e.g. Issue 123: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue #>` in the title with the actual Issue number.
> 
> ---

### CODE REVIEW

**Comment by sijie** [comment]
@reddycharan - can you review this change? since this is related to your change at #190 (9ddd9e6f9e48b03a57a2c78ec2630303abd49782)

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java [hunk 1]
```diff
@@ -398,7 +401,11 @@ synchronized long getLeastUnflushedLogId() {
     }
 
     synchronized long getCurrentLogId() {
-        return logChannel.getLogId();
+        return currentLogId;
+    }
+
+    BufferedLogChannel getCurrentLogChannel() {
```
[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: we can't just use `currentLogId` for testing the behavior of defer creation. defer creation means the `currentLogChannel` is null.

currentLogId and currentLogChannel are single entry log manager stuffs. they should not be part of the interface, that's what I suggested in our discussion, hiding all the log id stuffs in the implementation. If you follow that suggestion, the currentLogId and currentLogChannel should be just package-protected methods in single entry log manager, and the test EntryLogTest should be changed to test single entry log manager rather than entry logger.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java [hunk 2]
```diff
@@ -416,8 +423,10 @@ File getCurCompactionLogFile() {
     protected void initialize() throws IOException {
         // Register listener for disk full notifications.
         ledgerDirsManager.addLedgerDirsListener(getLedgerDirsListener());
-        // create a new log to write
-        createNewLog();
+
+        if (ledgerDirsManager.hasWritableLedgerDirs()) {
+            createNewLog();
```
[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: I know that. Your change involves a refactor which take time to get in, and probably we can't take it as 4.7.0 release. This change I am doing here is fixing a bug for 4.7.0 release. I don't think it is needed to couple these two changes together.

[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: I'm not sure about 4.7.0 release/timelines. But this PR - https://github.com/apache/bookkeeper/pull/1281 shouldn't take more time considering we agreed upon new interface. But anyhow with my change since I'm removing intialize method, the change you made here will be removed as well and since initialize method is removed, Bookie would be able to start with are without any writableLedgerDirs, as this change intend to be.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: my basic question is why do we need one more method - getCurrentLogChannel? why cannt we manage your testing needs with just one method here

```	 
synchronized long getCurrentLogId() {
    if (logChannel == null) {
      // return some invalid/unassigned id
    } else {
      return logChannel.getLogId();
    }
}
```

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: @reddycharan - I am starting the release 4.7.0 release this week. I don't think #1281 is able to make it to 4.7.0 since it is a bit *risky*. The change here is much simpler just to defer log creation.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: ideally EntryLogger isn't supposed to expose both currentLogId and currentLogChannel, but we made peace with the fact for currentLogId, but I'm not convenient with one more non-ideal decision.

[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: Current EntryLogger assumes it is single entry log implementation. These methods are package-protected methods (not public method), they are exposed for testing purpose.

----- 

If you are thinking from EntryLogManager perspective, these methods should be hidden to single entry log manager implementation. These methods should be only package-protected at single log manager implementation. 

If you see inconvenient here, that means your EntryLogManager interface in #1281 is not defining the right interfaces. `currentLogId` and `currentLogChannel` should only matter and be testing on single log entry manager. you shouldn't never consider it for per ledger entry manager. If you think and define your EntryLogManager interface in this way, I don't think `currentLogId` and `currentLogChannel` is not a problem to you.

[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: @reddycharan 

> my basic question is why do we need one more method - getCurrentLogChannel? why cannt we manage your testing needs with just one method here

why `getCurrentLogChannel` is a problem to you? I don't really understand your concern here. are you concerning the method making your interface complicate or what? 

as I said if you think this make your interface complicated, that mean your interface doesn't provide the right abstracted methods. The methods I am adding here is only for single entry log implementation, that says `currentLogId` and `currentLogChannel` should *ONLY* belong to single entry log implementation (which current EntryLogger is exactly implemented). 

After #1281, these two methods should *ONLY* moved to single entry log implementation. You shouldn't consider implement these two methods with PerLedgerEntryLogManager, otherwise that means your interface defines wrong methods.

[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: yeah these aren't public methods, but since there are packaged methods there is a possibility that classes in this package might use it, just like getcurrentlogid is called from SortedLedgerStorage, which I'm afraid of. For these classes, the type of the entrylogmanager should be transparent and at that time entrylogmanager for entrylogperledger should also provide that api.

But still I don't understand why do you need one more variable - currentLogId and one more method - getcurrentlogchannel, when your test cases are just checking if it is null or not null. It is extra burden to keep these vaiables in sync currentlogchannel and currentLogId and maintain one more method - getcurrentlogchannel (which I don't think should be exposed in the first place).

If you still feel strong about it, then let's have just currentLogChannel variable and getCurrentLogChannel method. We can remove currentLogId variable and it's corresponding getter. (for my entrylogmanager manager I would return null for this method).

[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: > just like getcurrentlogid is called from SortedLedgerStorage, 
> for my entrylogmanager for entrylogperledger I would return null for this method

as I pointed out at different places, log id stuffs are implementation specific and they should be hidden to implementation of EntryLogManager. they don't belong to the interface of EntryLogManager. I proposed in the other comment to have a method at EntryLogManager `write(SkipLIstFlusher ...)`, to write a skip list to entry logger, in this way you can hide the single-entrylog-manager implementation into this method without interfering with `getCurrentLogId`. so SortedLedgerStorage will not call `getCurrentLogId` any more but the tests can still examine `getCurrentLogId` through single-log-manager-implementation at white-box testing.

But again this is a discussion should be in #1281 rather than here.
 
> currentLogId and one more method - getcurrentlogchannel, when your test cases are just checking if it is null or not null. It is extra burden to keep these vaiables in sync currentlogchannel and currentLogId and maintain one more method 

That's a valid point to ask rather than from EntryLogManager perspective. I will remove `getCurrentLogChannel`.

[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java]: > That's a valid point to ask rather than from EntryLogManager perspective. I will remove getCurrentLogChannel.

Peace! thats my concern to begin with, having two variables and two getters when just one should be sufficient.

**Comment by sijie** [comment]
@reddycharan I removed the `getCurrentLogChannel` call.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/LedgerDirsManager.java [hunk 3]
```diff
@@ -176,19 +176,10 @@ public boolean hasWritableLedgerDirs() {
             return writableLedgerDirectories;
         }
 
-        // If Force GC is not allowed under no space
-        if (!forceGCAllowWhenNoSpace) {
```
Comment (reddycharan) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/LedgerDirsManager.java]: any reason for why this check is removed?

Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/LedgerDirsManager.java]: 1) `forceGCAllowWhenNoSpace` is controlling the gc behavior, it controls whether gc/compaction
2) this setting is replaced by `minUsableSizeForEntryLogCreation`. because not only gc/compaction will create the entry log files on readonly, during journal replays it would still need create entry log files. This allows we disableForceGC when no space but we can still allow bookie start up in readonly mode to replay journals if there is enough space.
